server/controllers/files.py

```diff
@@ -0,0 +1,57 @@
+import tempfile
+import logging
+
+import libcloud
+from flask import Blueprint, url_for, redirect, send_from_directory, abort
+from flask_login import login_required, current_user
+
+from server.extensions import storage
+from server.models import ExternalFile
+
+files = Blueprint('files', __name__)
+
+logger = logging.getLogger(__name__)
+
+@files.route("/files/<hashid:file_id>")
+@login_required
+def file_url(file_id):
+    ext_file = ExternalFile.query.filter_by(id=file_id, deleted=False).first()
+    if not ext_file or not ExternalFile.can(ext_file, current_user, 'download'):
+        return abort(404)
+    try:
+        storage_obj = ext_file.object
+    except libcloud.common.types.InvalidCredsError as e:
+        logger.warning("Could not get file {0} - {1}".format(file_id, ext_file.filename),
+                       exc_info=True)
+        storage_obj = None
+        raise e
+
+    if storage_obj is None:
+        return abort(404, "File does not exist")
+
+    # Do not use .download_url for local storage.
+    if "local" in storage.driver.name.lower():
+        return redirect(url_for('.send_file', file_id=ext_file.id))
+    else:
+        return redirect(storage_obj.download_url())
+
+@files.route("/files/<hashid:file_id>/download")
+@login_required
+def send_file(file_id):
```

Comment:
I don't really see why this is necessary. We shouldn't be downloading files to the web server machine in production. And to serve local files, we can just send the file directly instead of copying it somewhere else.